### ISSUE TITLE
Always specify parameterized type (to avoid implied dynamic)

### ISSUE THREAD

**Issue by awhitford** [dart-lang/linter#2339]
**Whenever one declares a parameterized type, the type must be specified.**
When a type is not specified, the parameterized type is `dynamic`; if the parameterized type needs to be `dynamic`, then be _explicit_ about it.

Consider code like this:
```dart
class Foo<T> {
  T _value;
  
  void saveValue(T value) {
    _value = value;
  }
}

main() {
  final foo1 = Foo<Int>(); // Good
  final foo2 = Foo(); // Bad, dynamic implied
  final foo3 = Foo<dynamic>(); // Good, dynamic explicit
}
```

While building a bunch of parameterized subclasses around [FormFieldState](https://api.flutter.dev/flutter/widgets/FormFieldState-class.html), there are a tonne of method like:
```dart
void didChange(T value)
```
I want the code to be passing around parameters with type `T`, not `dynamic`, because the stronger typing can uncover bugs at compile time instead of runtime.

**Additional context**

I tried using `always_specify_types`, but it disallows type inference for local variables, which is part of _Effective Dart_.

**Comment by srawlins** [comment]
Have you tried `strict-inference`? https://github.com/dart-lang/language/blob/master/resources/type-system/strict-inference.md

Speccifically under "Instance Creation", I believe this case is covered.

**Comment by bwilkerson** [comment]
@srawlins I believe that the `implicit-dynamic` option covers this case. Is that true?

**Comment by bwilkerson** [comment]
Sorry, you responded just before I clicked comment. Thanks!

**Comment by awhitford** [comment]
> Have you tried `strict-inference`? https://github.com/dart-lang/language/blob/master/resources/type-system/strict-inference.md
> 
> Specifically under "Instance Creation", I believe this case is covered.

@srawlins Thank you for this tip because it did highlight several opportunities, and it did capture my use-case above.

However, it failed to capture another use-case -- which I erroneously thought was the _same thing_ as above...

Please consider this code:
```dart
class Foo<T> {
  T _value;
  
  void saveValue(T value) {
    _value = value;
  }
}

void bar1<T>(Foo<T> foo) { // Good
  print ('${foo._value}');
}

void bar2<T>(Foo foo) { // Bad, dynamic implied
  print ('${foo._value}');
}

void foobar() {
  final foo1 = Foo<int>(); // Good
  bar1(foo1);
  bar2(foo1);
}
```

I'm interested in highlighting the missing `<T>` on `bar2`'s `foo` parameter.

**Comment by srawlins** [comment]
For the `Foo` found in bar2, I think you'll benefit from `strict-raw-types`: https://github.com/dart-lang/language/blob/master/resources/type-system/strict-raw-types.md

**Comment by awhitford** [comment]
> I believe that the [implicit-dynamic](https://dart.dev/guides/language/analysis-options#enabling-additional-type-checks) option covers this case. Is that true?

Thanks @bwilkerson, I added this:
```
  strong-mode:
    implicit-casts: false
    implicit-dynamic: false
```

And it does catch my refined scenario -- it highlights the call `bar2(foo1)` rather than the declaration `void bar2<T>(Foo foo)`, but I think that is OK.
